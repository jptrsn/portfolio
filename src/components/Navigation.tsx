'use client';

import { useState } from 'react';
import Link from 'next/link';
import { usePathname } from 'next/navigation';

const Navigation = () => {
  const pathname = usePathname();
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
  const [isAnimating, setIsAnimating] = useState(false);

  const toggleMobileMenu = () => {
    if (isMobileMenuOpen) {
      setIsAnimating(true);
      setTimeout(() => {
        setIsMobileMenuOpen(false);
        setIsAnimating(false);
      }, 300); // Match the transition duration
    } else {
      setIsMobileMenuOpen(true);
    }
  };

  const closeMobileMenu = () => {
    setIsAnimating(true);
    setTimeout(() => {
      setIsMobileMenuOpen(false);
      setIsAnimating(false);
    }, 300); // Match the transition duration
  };

  const navigationLinks = [
    { href: '/about', label: 'About' },
    { href: '/projects', label: 'Projects' },
    { href: '/experience', label: 'Experience' },
    { href: '/blog', label: 'Blog' },
    { href: '/contact', label: 'Contact' },
  ];

  const isActiveLink = (href: string) => {
    if (href === '/') {
      return pathname === '/';
    }
    return pathname.startsWith(href);
  };

  return (
    <>
      <nav className="fixed top-0 w-full z-50 bg-background/80 backdrop-blur-md border-b border-neutral-800">
        <div className="container mx-auto px-6 py-4 flex justify-between items-center">
          <Link href="/" className={`text-xl font-bold text-gradient ${isActiveLink('/') && 'text-gradient'}`}>
            EduCoder.dev
          </Link>

          {/* Desktop Navigation */}
          <div className="hidden md:flex space-x-8">
            {navigationLinks.map((link) => (
              <Link
                key={link.href}
                href={link.href}
                className={`${isActiveLink(link.href) ? 'text-secondary-500 transition-colors' : 'hover:text-primary-500 transition-colors'}`}
              >
                {link.label}
              </Link>
            ))}
          </div>

          {/* Mobile Hamburger Button */}
          <button
            className="md:hidden z-60 relative"
            onClick={toggleMobileMenu}
            aria-label="Toggle mobile menu"
            aria-expanded={isMobileMenuOpen || isAnimating}
          >
            <div className="w-6 h-6 flex flex-col justify-center space-y-1">
              <div
                className={`w-full h-0.5 bg-foreground transition-transform duration-300 ${
                  (isMobileMenuOpen || isAnimating) && !(!isMobileMenuOpen && isAnimating) ? 'rotate-45 translate-y-1.5' : ''
                }`}
              />
              <div
                className={`w-full h-0.5 bg-foreground transition-opacity duration-300 ${
                  (isMobileMenuOpen || isAnimating) && !(!isMobileMenuOpen && isAnimating) ? 'opacity-0' : ''
                }`}
              />
              <div
                className={`w-full h-0.5 bg-foreground transition-transform duration-300 ${
                  (isMobileMenuOpen || isAnimating) && !(!isMobileMenuOpen && isAnimating) ? '-rotate-45 -translate-y-1.5' : ''
                }`}
              />
            </div>
          </button>
        </div>
      </nav>

      {/* Mobile Menu Overlay */}
      <div
        className={`fixed inset-0 z-40 transition-opacity duration-300 md:hidden ${
          (isMobileMenuOpen || isAnimating) ? 'opacity-100 visible' : 'opacity-0 invisible'
        }`}
      >
        {/* Backdrop */}
        <div
          className="absolute inset-0 bg-black/50 backdrop-blur-sm"
          onClick={closeMobileMenu}
        />

        {/* Mobile Menu */}
        <div
          className={`absolute top-0 right-0 w-64 h-full bg-background border-l border-neutral-800 shadow-xl transform transition-transform duration-300 ${
            isMobileMenuOpen && !isAnimating ? 'translate-x-0' : 'translate-x-full'
          }`}
        >
          <div className="pt-20 px-6">
            <div className="flex flex-col space-y-6">
              {navigationLinks.map((link) => (
                <Link
                  key={link.href}
                  href={link.href}
                  className="text-lg hover:text-primary-500 transition-colors py-2 border-b border-neutral-800/50 last:border-b-0"
                  onClick={closeMobileMenu}
                >
                  {link.label}
                </Link>
              ))}
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default Navigation;