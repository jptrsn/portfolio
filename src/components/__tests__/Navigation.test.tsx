import { render } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import Navigation from '../Navigation'

describe('Navigation', () => {
  it('matches snapshot', () => {
    const { container } = render(<Navigation />)
    expect(container).toMatchSnapshot()
  })
})