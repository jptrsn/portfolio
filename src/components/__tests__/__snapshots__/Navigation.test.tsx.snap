// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation > matches snapshot 1`] = `
<div>
  <nav
    class="fixed top-0 w-full z-50 bg-background/60 backdrop-blur-md border-b border-neutral-800"
  >
    <div
      class="container mx-auto px-6 py-4 flex justify-between items-center"
    >
      <a
        href="/"
      >
        <span
          class="relative inline-block w-full"
        >
          <span
            aria-hidden="true"
            class="text-gradient invisible w-full inline-block"
          >
            EduCoder.dev
          </span>
          <span
            class="text-gradient absolute top-0 left-0 w-full"
            style="max-width: 100%; word-wrap: break-word;"
          />
          <style>
            
        .typewriter-cursor {
          animation: none;
        }

        @keyframes blink {
          0%, 50% {
            opacity: 1;
          }
          51%, 100% {
            opacity: 0;
          }
        }
      
          </style>
        </span>
      </a>
      <div
        class="hidden md:flex space-x-8"
      >
        <a
          href="/about"
        >
          About
        </a>
        <a
          href="/projects"
        >
          Projects
        </a>
        <a
          href="/posts"
        >
          Posts
        </a>
      </div>
      <button
        aria-expanded="false"
        aria-label="Toggle mobile menu"
        class="md:hidden z-60 relative"
      >
        <div
          class="w-6 h-6 flex flex-col justify-center space-y-1"
        >
          <div
            class="w-full h-0.5 bg-foreground transition-transform duration-300 "
          />
          <div
            class="w-full h-0.5 bg-foreground transition-opacity duration-300 "
          />
          <div
            class="w-full h-0.5 bg-foreground transition-transform duration-300 "
          />
        </div>
      </button>
    </div>
  </nav>
  <div
    class="fixed inset-0 z-40 transition-opacity duration-300 md:hidden opacity-0 invisible"
  >
    <div
      class="absolute inset-0 bg-black/50 backdrop-blur-sm"
    />
    <div
      class="absolute top-0 right-0 w-64 h-full bg-background border-l border-neutral-800 shadow-xl transform transition-transform duration-300 translate-x-full"
    >
      <div
        class="pt-20 px-6"
      >
        <div
          class="flex flex-col space-y-6"
        >
          <a
            href="/about"
          >
            About
          </a>
          <a
            href="/projects"
          >
            Projects
          </a>
          <a
            href="/posts"
          >
            Posts
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
